Improve no loader dialog

Pagwin2 opened the pull request:
Went and implemented what I suggested in issue #4369

Pagwin2 committed:
Implemented #4369

Signed-off-by: Pagwin <dev@pagwin.xyz>

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.h:
```diff
@@ -45,6 +45,8 @@
 class ModFolderPage : public ExternalResourcesPage {
     Q_OBJECT
 
+    friend void HandleNoModLoader(ModFolderPage* self);
```
make it a private function and use camelCase for function names

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -385,3 +387,34 @@ bool NilModFolderPage::shouldDisplay() const
+            if (self->m_instance->typeName() != "Minecraft") {
+                // not what we need
+                return;
+            }
```
safe to not check this

Pagwin2 replied:
Reason I did is due to noticing the check being present in some other code which got the instance and static casted to a `MinecraftInstance`. That said looking back at things yeah all places which call `HandleNoModLoader` check already so I'll remove the check

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -385,3 +387,34 @@ bool NilModFolderPage::shouldDisplay() const
+            auto profile = static_cast<MinecraftInstance*>(self->m_instance)->getPackProfile();
+            InstallLoaderDialog dialog(profile, QString(), self);
+            dialog.exec();
+            self->m_container->refreshContainer();
```
At the end, why not just open the dialog?

Pagwin2 replied:
Not sure if you're referring to the particular way I wrote this code or why I didn't open up the relevant followup dialog (download, update, or change mod version) 

For the former I basically copied the code from [launcher/ui/pages/instance/VersionPage.cpp:452](https://github.com/PrismLauncher/PrismLauncher/blob/8b43ea486baceed34501b0058857471b3c1bbfba/launcher/ui/pages/instance/VersionPage.cpp#L452) under the assumption that that was the best/only way to do it for some reason or another

For the latter laziness when I was writing this patch. That said now that I'm looking to see about doing that I'm not sure what the behavior should be for `ModFolderPage::updateMods` and `ModFolderPage::changeModVersion`. 
Both methods seem like they shouldn't be possible to call if a mod loader isn't already present.

Trial97 replied:
Yeah, but once you installed the loader with the dialo, what is stopping you from opening the dialog you wanted (download mods or change version dialog)

Pagwin2 committed:
Made changes to address feedback from code review (#4374)

Signed-off-by: Pagwin <dev@pagwin.xyz>

Pagwin2 commented:
Sorry for the spam of amending that one commit over and over again.

Ryex reviewed (approved):
:+1:

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -385,3 +389,31 @@ bool NilModFolderPage::shouldDisplay() const
+        }
+        default: {
+            // Unreachable
+            break;
```
I would still return true here

Trial97 replied:
```suggestion
            return true;
```

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.h:
```diff
@@ -45,6 +45,8 @@
 class ModFolderPage : public ExternalResourcesPage {
     Q_OBJECT
 
+    bool HandleNoModLoader();
```
```suggestion
    bool handleNoModLoader();
```

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -385,3 +389,31 @@ bool NilModFolderPage::shouldDisplay() const
 }
+
+// Helper function so this doesn't need to be duplicated 3 times
+inline bool ModFolderPage::HandleNoModLoader()
```
```suggestion
inline bool ModFolderPage::handleNoModLoader()
```

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -305,8 +308,9 @@ void ModFolderPage::changeModVersion()
     if (!profile->getModLoaders().has_value()) {
-        QMessageBox::critical(this, tr("Error"), tr("Please install a mod loader first!"));
-        return;
+        if (HandleNoModLoader()) {
```
```suggestion
        if (handleNoModLoader()) {
```

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -201,8 +203,9 @@ void ModFolderPage::updateMods(bool includeDeps)
     if (!profile->getModLoaders().has_value()) {
-        QMessageBox::critical(this, tr("Error"), tr("Please install a mod loader first!"));
-        return;
+        if (HandleNoModLoader()) {
```
```suggestion
        if (handleNoModLoader()) {
```

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -145,8 +146,9 @@ void ModFolderPage::downloadMods()
     if (!profile->getModLoaders().has_value()) {
-        QMessageBox::critical(this, tr("Error"), tr("Please install a mod loader first!"));
-        return;
+        if (HandleNoModLoader()) {
```
```suggestion
        if (handleNoModLoader()) {
```

Pagwin2 committed:
Minor remaining fixes for PR (#4374)

- changed case of `handleNoModLoader` to match project style

- made default case in switch statement for `handleNoModLoader` return
  true for safety

Signed-off-by: Pagwin <dev@pagwin.xyz>

Pagwin2 commented:
Made the remaining changes, everything good to merge?

DioEgizio commented:
if you first click yes on
<img width="788" height="386" alt="image" src="https://github.com/user-attachments/assets/bd1fd1c8-d420-42ba-8090-9c5fc2cddb67" />
but then you cancel in the install loader page, prism crashes

Trial97 commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -385,3 +389,32 @@ bool NilModFolderPage::shouldDisplay() const
+            // Should be safe
+            auto profile = static_cast<MinecraftInstance*>(this->m_instance)->getPackProfile();
+            InstallLoaderDialog dialog(profile, QString(), this);
+            dialog.exec();
```
return the result of exec from here and it should no longer crash

Pagwin2 replied:
Needed to return the negation of exec. Barring that this should be fixed now though

Pagwin2 committed:
Fixed Loader Install Cancel crash

Fixed a crash which occurred when the mod loader dialog was cancelled
after being reached from the no loader dialog

Signed-off-by: Pagwin <dev@pagwin.xyz>

getchoo commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -385,3 +389,32 @@ bool NilModFolderPage::shouldDisplay() const
+inline bool ModFolderPage::handleNoModLoader()
+{
+    int resp = QMessageBox::question(this, this->tr("Missing ModLoader"),
+                                     this->tr("You need to install a mod loader before installing mods, would you like to do so?"),
```
```suggestion
    int resp = QMessageBox::question(this, this->tr("Missing Mod Loader"),
                                     this->tr("You need to install a compatible mod loader before installing mods. Would you like to do so?"),
```

getchoo reviewed (commented):
Small nit, but LGTM. Thanks!

Pagwin2 committed:
Grammar/Spelling nit fix for Missing Mod loader dialog

Co-authored-by: Seth Flynn <getchoo@tuta.io>
Signed-off-by: Pagwin <git@pagwin.xyz>

Pagwin2 committed:
Edited a comment for clarity

Signed-off-by: Pagwin <git@pagwin.xyz>
